### PR TITLE
ARROW-3384: [Gandiva][C++] Sync remaining commits from Gandiva.

### PR DESCRIPTION
**Pull request by praveenbingo**
I've synced the rest of the commits from Gandiva.
Fixed all of the compilation and style warnings.

### CODE REVIEW

File: cpp/src/gandiva/llvm_generator_test.cc
Comment (wesm) [inline cpp/src/gandiva/llvm_generator_test.cc]: What are the 0's here?

File: java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java
Comment (wesm) [inline java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java]: Uses a private JIRA id

**Review by wesm — COMMENTED** [review]
Some minor comments.

**Comment by wesm** [comment]
To other maintainers -- leave this to me to merge since I want to tweak the commit messages a bit (add spaces after "[Gandiva]" for readability, edit the message on the patch with no "[Gandiva]")

File: cpp/src/gandiva/compiled_expr.h
Comment (praveenbingo) [reply cpp/src/gandiva/compiled_expr.h]: fixed..

File: cpp/src/gandiva/date_utils.cc
Comment (praveenbingo) [reply cpp/src/gandiva/date_utils.cc]: yes the pattern would be ascii only..

File: cpp/src/gandiva/expr_decomposer_test.cc
Comment (praveenbingo) [reply cpp/src/gandiva/expr_decomposer_test.cc]: looks like it it to simulate/test what happens when the visit method is called directly..

File: cpp/src/gandiva/filter_cache_key.h
Comment (praveenbingo) [reply cpp/src/gandiva/filter_cache_key.h]: sure done

File: cpp/src/gandiva/llvm_generator.cc
Comment (praveenbingo) [reply cpp/src/gandiva/llvm_generator.cc]: sure, changed that to a int64

File: cpp/src/gandiva/like_holder.h
Comment (praveenbingo) [reply cpp/src/gandiva/like_holder.h]: yes [https://github.com/google/re2/blob/master/re2/re2.h#L256](url)

File: cpp/src/gandiva/tests/projector_test.cc
Comment (praveenbingo) [reply cpp/src/gandiva/tests/projector_test.cc]: agreed, will fix as part of https://issues.apache.org/jira/browse/ARROW-3358

File: cpp/src/gandiva/llvm_generator_test.cc
Comment (praveenbingo) [reply cpp/src/gandiva/llvm_generator_test.cc]: added a comment.

File: java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java
Comment (praveenbingo) [reply java/gandiva/src/test/java/org/apache/arrow/gandiva/evaluator/ProjectorTest.java]: fixed.

File: cpp/src/gandiva/tests/if_expr_test.cc
Comment (praveenbingo) [reply cpp/src/gandiva/tests/if_expr_test.cc]: agreed,we are already doing a story to make it easier to generate an expression from a string, we can do something similar for data..

File: cpp/src/gandiva/projector_cache_key.h
Comment (praveenbingo) [reply cpp/src/gandiva/projector_cache_key.h]: as of now no, same expression with different names will be a new compilation.

opening a issue to investigate more.

File: cpp/src/gandiva/date_utils.cc
Comment (praveenbingo) [reply cpp/src/gandiva/date_utils.cc]: fixed.

File: cpp/src/gandiva/filter_cache_key.h
Comment (wesm) [inline cpp/src/gandiva/filter_cache_key.h]: Surprised this does not cause a compiler warning

File: cpp/src/gandiva/projector_cache_key.h
Comment (wesm) [inline cpp/src/gandiva/projector_cache_key.h]: Same re: compiler warning

**Review by wesm — APPROVED** [review]
+1. I'll merge after the build runs

**Comment by wesm** [comment]
Just normalized the commit messages and added spaces for readability

**Comment by wesm** [comment]
+1

**Comment by wesm** [comment]
Merged. Thank you!

**Pull request closed without merging**
